README
____         _____
|   \       /
|___/ \   / \____   __          __
|      \ /       \ /__\ |/ | / /__\ |/
|       /   _____/ \__  |  |/  \__  |
      _/

-----------------------------------
Server HTTP basato su Python
Copyright (c) 2014 Meneghetti Fabio
-----------------------------------

Introduzione
-----------------------------------
PyServer è un server HTTP/HTTPS leggero, scritto in python cercando di utilizzare
meno moduli esterni possibili

Utilizzo
-----------------------------------
Per utilizzare il server basta eseguire "main.py" con un interprete python
nella cartella dove è presente

Requisiti
-----------------------------------
* Python versione 2.6/2.7
* I moduli socket, time, os, sys, mimetypes, zlib, re
* Il modulo OpenSSL per il server HTTPS

Configurazione
-----------------------------------
In questa versione NON è più necessario modificare il file config.py
La prima variabile nel file 'main.py' indica un file di configurazione.
Sebbene non sia necessario è possibile crearlo e modificare a piacere gli attributi sotto,
rispettando la sintassi conf e le sezioni. Tutto ciò che non viene menzionato o modificato
viene lasciato con il suo valore di default
Nota: Se il modulo OpenSSL non dovesse essere disponibile (Es. su Android)
il server sicuro verrà disabilitato anche se abilitato nella configurazione

Tra parentesi quadre è indicato il valore originale
Se il campo è racchiuso tra parentesi graffe vuol dire che la feature è
in sviluppo e non funzionerà indipendentemente dal settaggio

la variabile cfg è la configurazione cartella relativa al file

Sezione "debug"
level        Imposta il livello di debug
             (D, I, W, E; rispettivamente debug, info, attenzione, errore)
             Un qualunque altro valore (Anche le lettere sopra in minuscolo)
             disabilita il debug
             ["I"]
             
use_colors   Su sistemi unix/linux o comunque terminali
             che supportano lo standard vt100 è possibile colorare
             i tag di debug per distinguerli meglio
             [False]
             
file_append  Se impostato a True aggiunge il debug al file invece che
             sovrascrivere il file stesso
             [False]


Sezione "paths"
debug       E' possibile scrivere il debug anche su file.
            Per disabilitarlo lasciare il campo vuoto
            ["PyDebug.log"]
            
webdisk     La stringa indica un percorso, relativo o assoluto, indicante la radice
            dei file disponibili sul server
            ["webdisk"]
            
folder_cfg  A discapito del nome (mi girava così in quel momento) contiene le configurazioni delle cartelle dell' HD virtuale
            E' un dizionario a due strati (Pagine -> configurazione relativa alla stessa): tuttavia l'unico accessibile
            direttamente in pyml è l'ultimo strato (tramite 'configuration')
            Il file si presenta come un classico conf e questo è il nome che avrà per ogni cartella,
            dove le categorie corrispondono ai file a cui è riferita la configurazione.
            Le categorie funzionano anche come due tipi di regex:
              la prima (shell type) scrivendola semplicemente nelle parentesi quadre e supporta i selettori '*' e '?' (anche multipli) 
                Es: [*.mk?] che seleziona (tra gli altri) tutti i file mkv e tutti i file mka
              la seconda (regex) scrivendo nelle parentesi quadre 're:' seguito dalla regex, SENZA SPAZIH!
                Es: [re:.*\.(bmp|jpe?g|gif|png)] che seleziona (quasi) tutti i comuni file immagine (bmp, jpg, jpeg, gif, png)
            
cache       Qui verranno salvate le copie cache in formato GZip modificato (4 bytes per il secondo di scadenza e il resto GZip)
            ["tmp"]
            
https_key   La chiave privata e il certificato per il server HTTPS
https_cert  Su linux/unix si possono generare con il comando
                openssl genrsa -des3 -out chiave_privata_server.key 1024
                openssl req -new -key chiave_privata_server.key -out chiave_privata_server.cert
            Verranno generati i file chiave_privata_server.key e
            chiave_privata_server.cert con algoritmo des3
            [""], [""]


Sezione "buffers"

post_undefined_length  Quando il client esegue una richiesta POST senza il campo Content-Length nell'header
                       questa è la dimensione di ricezione del buffer
                       [128]
                       
send_buffer            Quando i file vengono inviati al client questa è la dimensione del buffer
                       Non deve mai essere superiore alla Window Size TCP (Generalmente 65536 Bytes)
                       [2048]
                       
tcp_receive_buffer     Questa è la dimensione di ricezione dei dati nelle richieste POST
                       Non deve mai essere superiore alla Window Size TCP (Generalmente 65536 Bytes)
                       [32768]


Sezione "cache"

use                   Abilita la cache. Non può essere bypassato dalla cfg
                      [True]
                      
on_pages              Abilita il caching delle pagine statiche. Può essere bypassata. Valida solo su file .html e .htm
                      [True]
                      
pages_validity        Validità della cache delle pagine dalla creazione (o aggiornamento) espresso in secondi. Può essere bypassato dalla cfg
                      [300]
                      
on_files              Abilita il caching dei file. Può essere bypassata. Non valida su pagine statiche e dinamiche
                      [True]
                      
files_validity        Validità della cache dei file dalla creazione (o aggiornamento) espresso in secondi. Può essere bypassato dalla cfg
                      [300]
                      
on_large_files        Abilita il caching dei file grandi. Può essere bypassata. Non valida su pagine statiche e dinamiche
                      [True]
                      
large_files_low_size  Dimensione sopra la quale un file viene considerato grande. Non può essere bypassata
                      [1024*1024*4 alias 4 MiB]
                      
lfiles_validity       Validità della cache dei file dalla creazione (o aggiornamento) espresso in secondi. Può essere bypassato dalla cfg
                      [600]


Sezione "misc"
index_search_order    La lista indica l'ordine di ricerca del file di index
                      quando un client richiede una pagina tipo http(s)://indirizzoserver/ o
                      http(s)://indirizzoserver/cartelle/altre_cartelle/
                      Se nessun file viene trovato viene inviato un errore 404
                      [[".pyml", ".pyhtml", ".html", ".htm"]] (è una lista!)
                      
server_error_template La stringa indica un file da cui verrà estratto il template di errore del client/server
                      Il file è un html standard, dove '%errnum' e '%errdesc' vengono rispettivamente sostituite
                      con il codice errore HTTP e una breve stringa di descrizione
                      
{chroot_after_load}   Se impostato a True, una volta completato il caricamento dei server il programma
                      esegue un chroot su WEBDISK_PATH in modo da aumentare la sicurezza.
                      Richiede i privilegi di amministratore/root
                      Linux/Unix: Cambia anche i permessi dell'utente a daemon dopo
                                  chroot e dopo aver modificato
                                  ad a+rwx i permessi di paths.webdisk
                                  Prima della chiusura del programma vengono reimpostati i permessi
                      [False]


Sezione "http"
use      Abilita il server HTTP
         [True]

port     La porta per il server HTTP. Per accedere al server usare http://indirizzoserver:porta/
         Se la porta è la 80 non serve aggiungere :porta all'URL, ma richiede i
         privilegi di root/amministratore per l'esecuzione del server
         [8080]
         
backlog  La coda del server HTTP. Un valore grande diminuisce il rischio
         di DoS (in particolare SYN Flood)
         a scapito delle prestazioni di rete
         [256]


Sezione "https":
use      Se impostato a True viene attivato anche il server HTTPS.
         Viene ignorata se il modulo OpenSSL non è presente.
         Sono inoltre necessari una chiave privata e un certificato
         (Vedi paths.https_key e paths.https_cert)
         [False]
         
port     La porta per il server HTTPS.
         Per accedere al server usare https://indirizzoserver:porta/
         Se la porta è la 443 non serve aggiungere :porta all'URL, ma richiede i
         privilegi di root/amministratore per l'esecuzione del server.
         [8443]
         
backlog  La coda del server HTTP. Un valore grande diminuisce il rischio
         di DoS (in particolare SYN Flood)
         a scapito delle prestazioni di rete
         [128]



Python Markup Language
-----------------------------------
Il server mette a disposizione un linguaggio ibrido tra HTML e Python,
con estensione .pyhtml o .pyml
eseguito lato server e inviato poi al client

Per inserire codice python in un pyml lo si deve includere nei tag <?py e ?>
    <?py
        print "ciao <br />"
        print "Mondo!<br />"
    ?>
Si può inserire codice inline mettendo i tag sulla stessa linea,
ma possono contenere una sola istruzione (o più se separate da ";")
    <title><?py print "Titolo dinamico!" ?></title>

Tutte le variabili create in un blocco sono disponibili in tutti quelli successivi

Un po di variabili servono ad integrare python con il server:
'document' è una classe che contiene:
    url: l'url grezzo
    args: lista argomenti passati nell'url (file?arg1=val1&arg2=val2...)
    header: contiene l'header di richiesta
    request: il tipo di richiesta (GET o POST)
    post: Se la richiesta è post, contiene tutti gli argomenti passati dal browser, separati come l'url
    rawpost: contiene i dati grezzi ricevuti dal client durante il POST,
             utili per utilizzare XMLHttpRequest o inviare file dal client (anche se conviene bypassare il post
                                                                            e riceverlo manualmente in questo caso)

'stream' è una classe che contiene:
    socket: il socket collegato al client
    header: l'header di risposta
    metodi di compressione, modificabili con .setCompressMode(tipo, qualità)
    sostituzione di '\n' in '<br />', abilitato di default, modificabile con .set_n_as_br(modo)
    streaming live, disabilitato di default, modificabile con .setWait(modo)
    flush coda, con .flush()
    

Gli argomenti e i dati POST sono automaticamente convertiti dal formato application/x-www-form-urlencoding ad ASCII
o lasciati così come sono se vengono rilevati come binari

Inoltre 'args' e 'post' sono dizionari particolari: infatti per accedere al valore di 'argomentò nell'url

    http(s)://indirizzo:porta/percorso/file?argomento=ciao

si può fare usandolo come chiave: document.args["argomento"]
oppure usandolo come valore:      document.args.argomento

L'header di risposta è modificabile in stile quasi PHP (Forse è meglio questo modo =)
    stream.header(campo)         ritorna il valore di 'campo'
    stream.header(campo, valore) setta campo a valore e ritorna valore

I valori numerici e password NON SONO CONVERTITI AUTOMATICAMENTE!

Folder Cfg
-----------------------------------
Il server mette a disposizione la possibilità di creare un file per cartella in grado di modificare il comportamento
di uno o più file. DEVONO essere chiamati come config.paths.folder_cfg (predefinito con "." all'inizio, per nasconderlo).
La struttura del file è il solito conf, dove le categorie sono i nomi dei file (possono essere anche shell regex o normali regex anteponedo 're:')
I campi sono:
    receive_post              Ricevi il post in modo automatico. Serve per pagine dinamiche personalizzate [True]
    force_cache_state         Forza l'utilizzo o meno della cache. 1 forza l'uso, -1 forza il non uso
                              e 0 lascia decidere alla configurazione server [0]
    force_compression         Forza un metodo di compressione. Sono supportati solo identity, deflate e gzip.
                              Una stringa vuota lascia la decisione automatica [""]
    force_compression_quality Forza la qualità delle stringhe deflate, da 0 (Niente compressione, velocità alta)
                              a 9(Compressione massima, velocità bassa) [5]
    force_cache_validity      Forza la validità della cache a n secondi. 0 lascia l'impostazione automatica [0]

Minishell
-----------------------------------
Il server mette a disposizione una minishell contenenti alcuni comandi
per il debug sia del server, sia delle pagine dinamiche (/ separa le alternative):
    clients/c       Visualizza lo stato dei client attualmente collegati
    disconn/d       Disconnetti tutti i client troncando le connessioni correnti
    http/https      Seleziona il server HTTP/S. I comandi precedenti saranno validi solo su questo server
    q/quit/exit     Interrompi il server. Dato l'utilizzo di SO_REUSEADDR sul socket del server è possibile
                    reinizializzare il server subito sulla stessa porta, invece che aspettare il TCP Timeout


CHANGELOG
2014-05-14
    * pyml.py: (PyML) Aggiunta variabile self.script_wd per PyML.exec_segment
    * pyml.py: (PyML.exec_segment) Rimosso bug della WorkingDirectory degli script (wd relativa alla radice dei file)
2014-05-15
    * main.py: (Client.start) Ricostruito metodo di individuazione del tipo di richiesta
                              Aggiunta variabile self.request, contenente la richiesta HTTP
                              Aggiunto supporto per richieste HEAD
    * main.py: (Client.elab) Aggiunto supporto richiesta HEAD
    * main.py: (Client.sendError) Aggiunto supporto richiesta HEAD
2014-05-22
    * main.py: (Client.start) Assegnati i metodi HTTP PUT, DELETE, PATCH, TRACK, OPTIONS all'errore 405
    * main.py: (Client.sendError) Aggiunto errore 405 Method Not Allowed
2014-06-11
    * main.py: (Client.prepareFile) Aggiunta differenziazione metodi 200 OK e 206 Partial Content
    * main.py: (Client.start) Aggiunto differenziazione metodi 200 e 206 per lo streaming
2014-06-13
    * changelog: Sistemato un typo :)
                 Cambiato il nome in CHANGELOG (Conforme agli standard)
    * pyml.py (PyML.exec_segment): Aggiunto supporto alle sequenze di escape ascii, esadecimali e ottali
    * shared.py (decodeurl): Aggiunto riconoscimento stringhe binarie
    Prima release su github.com (https://github.com/FabioMene/PyServer)
2014-06-26 -> 2014-06-29 (Ehhhhh!)
    * Aggiornato il progetto a versione 2.3.0
    * main.py: Aggiunta variabile 'CONFIG_FILE' per il file di configurazione
               Configurazione completamente rielaborata, in formato conf
               Controllo configurazione e accessibilita' cartelle all'avvio
               Cambiato il tag http da 'Serv' a 'HTTP '
               Cambiato il tag https da 'SSL' a 'HTTPS'
               (ClientInit) Globalizzato metodo risoluzione traceback (inserito in shared.py)
               (File) Cambiata nome classe in 'Page' e rimosso l'attributo 'mime' (text/html)
               (Client) Aggiunta variabile 'compression'
               (Client.elab) Aggiunto metodo di selezione della compressione, in ordine: gzip, deflate, identity
                             Aggiunta variabile 'cfg' (dopo aver verificato l'esistenza del file) contenente la configurazione della cartella
                             Modificato metodo  di elaborazione PyMl a dinamico (finalmente :D)
                             Aggiunto campo 'Content-Encoding' nella sezione di streaming file
               (Client.preparePage) Aggiunta elaborazione metodi di compressione e di caching, tenendo conto di tutti i modificatori in 'cfg'
                                    Aggiunta elaborazione del traceback in caso di errore (Si spera di no ;D)
               (Client.prepareFile) Aggiunta elaborazione metodi di compressione e di caching, tenendo conto di tutti i modificatori in 'cfg'
                                    Aggiunta elaborazione del traceback in caso di errore (Si spera di no ;D) (Copia e incolla evidente ...)
                                    NOTAH! le richieste contenenti 'Range' saranno trattate comunque come 200 OK, tranne nel caso che la compressione si identity
               Cambiato qualcosa a livello di selezione server nella minishell
    * pyml.py: (Document) Aggiunto header di richiesta come 'document.header' accessibile nel solito modo (get=header(attr), set=header(attr, val))
               (Stream) Aggiunta classe Stream, contente i dettagli di streaming quali socket(client), header e altra robba
                        Supporto alla compressione on-the-fly (o al volo per tradizionalisti) e cambio dei '\n' a '<br />' (Conforme al w3c, mica medda!) [Predef: abilitato]
                        Supporto al live streaming con stream.setWait(False) o all'accodamento di cio' che viene inviato con stream.setWait(True)
                        Classe inizializzata richiamabile tramite 'stream' o 'sys.stdout' (Si' e' un piccolo hack di sys =)
                        NOTAHH! il metodo finish viene chiamato dall'esterno, non e' possibile chiamarlo nello script perche' genera alla fine del'esecuzione della pagina un errore 
               (PyMl) Aggiunto override dei metodi 'exit' e 'quit' che terminano e inviano la pagina al client ISTANTANEAMENTEH!
    * shared.py: Aggiunto un fracasso di robba, tra cui configurazioni, compressioni e bombe al plasma
                 (Dict) rifatto il metodo di richiesta elementi
                 (ServerConfiguration) Classe (inizializzata globalmente come 'config') per leggere la configurazione del server.
                                       Non e' tollerante: se un qualsiasi attributo o tipo di valore sono errati la configurazione viene ripristinata al default
                 (DEFAULT_ATTRIBUTES) Questo e' un dizionario contenente i default della configurazione delle cartelle (Simile ad .htaccess ma (forse) piu' esteso)
                                      Ad ogni chiave corrisponde una lista di due elementi: rispettivamente tipo e valore di default
                 (FileConfiguration) Contiene la configurazione cartella
                 (.*) Spostati qui tutti i metodi relativi alla generazione di header HTTP(S)
                 (Compressor) Fornisce una classe per la compressione DEFLATE + il calcolo di dimesione e crc32
                 (get_gzip) Crea uno stream GZip tramite Compressor, ritornato on-the-fly come stringa (Con header e crc32)
    * config.py: ATTENZIONEH! Questo file non e' piu' la configurazione del server! Modifica la variabile CONFIG_FILE in main.py per designare un file di configurazione personalizzato
                 Nel caso in cui:
                    - il file di configurazione non sia presente
                    - la configurazione sia buggata (bacata, corrotta, chiamala come ti pare ;)
                    - si verifichi un attacco alieno

pyml.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#  The MIT License (MIT)
#  
#  Copyright (c) 2014 Meneghetti Fabio
#  
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#  
#  The above copyright notice and this permission notice shall be included in all
#  copies or substantial portions of the Software.
#  
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#  SOFTWARE.

# Parser PyMl

import os, sys, shared, copy, struct, zlib, time
from multistream import NullFile

config={}

class Document(shared.Dict):
    def __init__(self, page, req, args, post, hdr):
        shared.Dict.__init__(self)
        self.request=req
        self.url=page+"?"+args
        if args=="": args={}
        else:
            t={}
            for s in args.split("&"):
                v=s.split("=")
                if len(v)==1: t[v[0]]=""
                else: t[v[0]]=shared.decodeurl(v[1])
            args=t
        self.args=shared.Dict(args)
        if not post:
            self.rawpost=""
            post={}
        else:
            t={}
            self.rawpost=post
            for s in post.split("&"):
                v=s.split("=")
                if len(v)==1: t[v[0]]=""
                else: t[v[0]]=shared.decodeurl(v[1])
            post=t
        self.post=shared.Dict(post)
        self.header=hdr

class Stream(shared.Dict):
    def __init__(self, client, req):
        shared.Dict.__init__(self)
        self.__request=req
        self.socket=client
        self.__wait=True
        self.queued_data=""
        self.queued_data_size=0
        self.__comptype="identity"
        self.__compdata=None
        self.header=shared.Header({"Content-Encoding": "identity", "Content-Type": "text/html;charset=utf-8"})
        self.header.request="200 OK"
        self.n_is_br=True
        self.__hdrsent=False
    def write_plain(self, data, final=False):
        segm=self.__compress(data, final)
        self.__send(segm)
    def write(self, data, final=False):
        if not data:
            self.__send("")
            return 0
        if data[-1] == "\n": data=data[:-1]
        segm=""
        i, l=0, len(data)
        while i < l:
            add=1
            if data[i]=="\\":
                add=2
                if i == l-1:
                    return "Errore(PyMl Parser): Sequenza \"\\\" non terminata<br />"
                elif data[i+1] == "\n": pass
                elif data[i+1] == "n": segm+="\n"
                elif data[i+1] == "t": segm+="\t"
                elif data[i+1] == "\\": segm+="\\"
                elif data[i+1] == "\"": segm+="\""
                elif data[i+1] == "\'": segm+="\'"
                elif data[i+1] == "x":
                    if i+4 >= l:
                        return "Errore(PyMl Parser): Sequenza esadecimale \"\\xnn\" non terminata<br />"
                    try:
                        segm+=chr(int(data[i+2:i+4], 16))
                        add=4
                    except:
                        return "Errore(PyMl Parser): Sequenza esadecimale \"\\x%s\" non valida<br />"%(data[i+2:i+4])
                elif data[i+1] >= "0" and data[i+1] <= "9":
                    if i+4 >= l:
                        return "Errore(PyMl Parser): Sequenza ottale \"\\nnn\" non terminata<br />"
                    try:
                        segm+=chr(int(data[i+1:i+4], 8))
                        add=4
                    except:
                        return "Errore(PyMl Parser): Sequenza ottale \"\\%s\" non valida<br />"%(data[i+1:i+4])
                else:
                    return "Errore(PyMl Parser): Sequenza \"%s\" non valida<br />"%(data[i+1])
            else: segm+=data[i]
            i+=add
        if self.n_is_br: segm=segm.replace("\n", "<br />")
        l=len(segm)
        segm=self.__compress(segm, final)
        self.__send(segm)
        return l
    def __send(self, data):
        if self.__wait:
            self.queued_data+=data
            self.queued_data_size+=len(data)
        else:
            if not self.__hdrsent:
                if not self.header("Date"): self.header("Date", shared.generateDate()[6:-2])
                self.socket.send(self.header.build())
                self.__hdrsent=True
                if self.__comptype == "gzip" and self.__request != "HEAD":
                    self.socket.send("\x1f\x8b\x08\x00"+struct.pack("<L", long(time.time()))+"\x00\xff") #GZIP Header (deflate, unknown, mtime)
            if self.__request != "HEAD":
                while self.queued_data:
                    self.socket.send(self.queued_data[:config.buffers.send_buffer])
                    self.queued_data=self.queued_data[config.buffers.send_buffer:]
                self.queued_data=""
                self.queued_data_size=0
                while data:
                    self.socket.send(data[:config.buffers.send_buffer])
                    data=data[config.buffers.send_buffer:]
    def setCompressMode(self, comp, quality=5):
        if self.__hdrsent: return
        self.__comptype=comp
        self.__compdata=None
        if   comp == "identity": pass
        elif comp == "deflate":
            self.__compdata=zlib.compressobj(quality)
        elif comp == "gzip":
            self.__compdata=[zlib.compressobj(quality, zlib.DEFLATED, -zlib.MAX_WBITS, zlib.DEF_MEM_LEVEL, 0), 0, 0]
        else:
            raise ValueError("Tipo di compressione '%s' non supportata"%comp)
        self.header("Content-Encoding", comp)
    def __compress(self, data, final=False):
        if self.__comptype   == "identity": return data
        elif self.__comptype == "deflate":
            out=self.__compdata.compress(data)
            if final:
                out+=self.__compdata.flush(zlib.Z_FINISH)
            return out
        elif self.__comptype == "gzip":
            out=self.__compdata[0].compress(data)
            self.__compdata[1]+=len(data)
            self.__compdata[2]=zlib.crc32(data, self.__compdata[2]) & 0xffffffffL
            if final:
                out+=self.__compdata[0].flush(zlib.Z_FINISH)
                out+=struct.pack("<LL", self.__compdata[2], self.__compdata[1] & 0xffffffffL)
            return out
    def setWait(self, mode):
        self.__wait=mode
    def flush(self):
        if self.__comptype   == "identity": return
        elif self.__comptype == "deflate":
            out=self.__compdata.flush(zlib.Z_SYNC_FLUSH)
        elif self.__comptype == "gzip":
            out=self.__compdata[0].flush(zlib.Z_SYNC_FLUSH)
        self.__send(out)
    def finish(self):
        self.__send(self.__compress("", True))
    #~ def setCharset(self, ch):
        #~ try:
            #~ test="test".encode(ch)
        #~ except:
            #~ raise ValueError("Codifica '%s' non supportata"%ch)
        #~ hdr=self.header("Content-Type").split(";")
        #~ m=False
        #~ for i in range(len(hdr)):
            #~ if "charset=" in hdr[i]:
                #~ hdr[i]="charset="+ch
                #~ m=True
        #~ if not m: hdr.append("charset="+ch)
        #~ self.header("Content-Type", ";".join(hdr))
    def set_n_is_br(self, mode): self.n_is_br=mode

class PyMl_Exit_Request(BaseException):
    def __init__(self):
        pass

class PyMl:
    def __init__(self, file, req, args, post, inheader, client, configuration):
        self.env={}
        self.document=Document(file, req, args, post, inheader)
        self.stream=Stream(client, req)
        
        null=NullFile()
        sys.stdout.registerThread(self.stream)
        sys.stderr.registerThread(self.stream)
        sys.stdin.registerThread(null)
        
        self.stream.header("Location", inheader("Host") or "")
        if "gzip" in inheader("Accept-Encoding"): self.stream.setCompressMode("gzip")
        elif "deflate" in inheader("Accept-Encoding"): self.stream.setCompressMode("deflate")
        else: self.stream.setCompressMode("identity")
        self.configuration=configuration
        self.script_wd=os.path.split(file)[0]
        f=open(file)
        cont=f.readlines()
        f.close()
        for x in range(len(cont)):
            cont[x]=cont[x].replace("\n", "").replace("\r", "")
        segm=[]
        self.line=1
        self.running=True
        for l in cont:
            if not self.running:
                break
            if "<?py" in l and "?>" in l:
                before, expr, after=self.get_expression(l)
                self.stream.write_plain(before)
                self.exec_segment(expr)
                self.stream.write_plain(after)
            elif "<?py" in l:
                segm=[l]
            elif "?>" in l:
                segm.append(l)
                segm="\n".join(segm)
                before, expr, after=self.get_expression(segm)
                self.stream.write_plain(before)
                self.exec_segment(expr)
                self.stream.write_plain(after)
                segm=[]
            elif segm!=[]: segm.append(l)
            else: self.stream.write_plain(l+"\n")
            self.line+=1
        self.stream.finish()
        self.stream.header("Content-Length", self.stream.queued_data_size)
        self.stream.setWait(False)
        self.stream.write("", True)
        sys.stdout.unregisterThread()
        sys.stderr.unregisterThread()
        sys.stdin.unregisterThread()
    def get_expression(self, expr):
        bf=ex=af=""
        for x in range(len(expr)):
            bf+=expr[x]
            if bf[-4:]=="<?py": break
        if bf==expr: return (expr, "", "")
        bf=bf[:-4]
        for x in range(x+1, len(expr)):
            ex+=expr[x]
            if ex[-2:]=="?>": break
        if bf+ex==expr: return (bf, ex, "")
        return (bf, ex[:-2], expr[x+1:])
    def exec_segment(self, segm):
        segm=self.indentify(segm)
        cwd=os.getcwd()
        os.chdir(self.script_wd)
        argv=copy.deepcopy(sys.argv)
        sys.argv=[]
        err=""
        try:
            self.execute(segm)
        except BaseException as e:
            if type(e) == PyMl_Exit_Request:
                self.running=False
            else:
                err=str(e)
        os.chdir(cwd)
        sys.argv=copy.deepcopy(argv)
        if err: self.stream.write_plain("Errore(PyMl Parser): %d: %s<br />"%(self.line, err))
        import sys as private_sys
        quit=exit=private_sys.exit
        del private_sys
    def exit(self):
        raise PyMl_Exit_Request()
    def execute(donotdelete, segmenttonotdelete):
        document=donotdelete.document
        configuration=donotdelete.configuration
        stream=donotdelete.stream
        for x in donotdelete.env:
            exec("%s = donotdelete.env[\"%s\"]"%(x, x))
        quit=donotdelete.exit
        exit=donotdelete.exit
        # # # # # # # # # # # # #
        exec(segmenttonotdelete)
        # # # # # # # # # # # # #
        for x in dir():
            if x=="x" or x=="donotdelete" or x=="segmenttonotdelete" or x=="outtonotdelete" or x=="document" or x=="stream" or x=="quit" or x=="exit": pass
            else: exec("donotdelete.env[\"%s\"] = %s"%(x, x))
        donotdelete.document=document
        donotdelete.stream=stream
    def indentify(self, segm):
        out=""
        ind=0
        for x in segm:
            if x!="\n": break
            segm=segm[1:]
        for x in segm:
            if x!=" ": break
            ind+=1
        for x in segm.split("\n"):
            out+=x[ind:]+"\n"
        return out


config.py
#  The MIT License (MIT)
#  
#  Copyright (c) 2014 Meneghetti Fabio
#  
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#  
#  The above copyright notice and this permission notice shall be included in all
#  copies or substantial portions of the Software.
#  
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#  SOFTWARE.

# Configurazione di default

DEFAULT_CONFIG={
    # debug
    "debug": {
        "level":       "I",   # livello di debug: D (Debug), I (Info), W (Attenzione), E (Errore)
        "use_colors":  False, # Utilizza colorazione livelli su terminali vt100 (o compatibili)
        "file_append": False  # Accoda il debug al file (config.paths.debug) invece che sovrascriverlo
    },

    #Percorso file
    "paths": {
        "debug":      "PyDebug.log",                   # File di debug
        "webdisk":    "webdisk",                       # HD Virtuale
        "folder_cfg": ".pyserver.folder_options.conf", # nome dei file di configurazione cartella
        "cache":      "tmp",                           # Posizione copie cache
        "https_key":  "",                              # Chiave privata SSL
        "https_cert": ""                               # Certificato pubblico SSL
    },

    #Comportamento alla mancanza di Content-Length in POST
    "buffers": {
        "post_undefined_length": 128,  # Content-Length in POST non definito
        "send_buffer":           2048, # Buffer di invio
        "tcp_receive_buffer":    32768 # Buffer di ricezione (Non dovrebbe essere superiore a 65536)
    },

    #Caching pagine e file
    "cache": {
        "use":                  True,        # Abilita cache
        "on_pages":             True,        # Abilita caching delle pagine. Puo' essere bypassato con opt->force_cache_state
        "pages_validity":       300,         # Validita' delle copie cache delle pagine in secondi dalla creazione
        "on_files":             True,        # Abilita caching dei file. Puo' essere bypassato con opt->force_cache_state
        "files_validity":       300,         # Validita' delle copie cache dei file in secondi dalla creazione
        "on_large_files":       False,       # Abilita caching dei file grandi. Puo' essere bypassato con opt->force_cache_state
        "large_files_low_size": 1024*1024*4, # Dimensione sopra la quale un file viene considerato grande
        "lfiles_validity":      600,         # Validita' delle copie cache dei file grandi in secondi dalla creazione
    },

    #Un po di misc non fa mai male!
    "misc": {
        "index_search_order":    [".pyml", ".pyhtml", ".htm", ".html"], # Ordine di ricerca del file index
        "server_error_template": "" # Maschera errori HTTP
    },

    #Configurazione porte
    "http": {
        "use":     True, # Usa HTTP
        "port":    8080, # Porta HTTP
        "backlog": 256   # Coda di accettazione
    },

    "https": {
        "use":     False, # Usa HTTPS
        "port":    8443,  # Porta HTTPS
        "backlog": 128    # Coda di accettazione
    }
}
